features/core/src/main/java/org/apache/karaf/features/internal/service/BundleInstallSupportImpl.java

```diff
@@ -140,13 +140,17 @@ public Bundle installBundle(String region, String uri, InputStream is) throws Bu
 
     @Override
     public void updateBundle(Bundle bundle, String uri, InputStream is) throws BundleException {
+        File file = null;
         // We need to wrap the bundle to insert a Bundle-UpdateLocation header
         try {
-            File file = BundleUtils.fixBundleWithUpdateLocation(is, uri);
+            file = BundleUtils.fixBundleWithUpdateLocation(is, uri);
             bundle.update(new FileInputStream(file));
-            file.delete();
         } catch (IOException e) {
             throw new BundleException("Unable to update bundle", e);
+        } finally {
+            if (file != null && file.exists()) {
```

Comment:
Just a minor comment while the checks run — I don’t think there’s much point in calling `.exists()` before `.delete()` (there’s still a TOCTTOU race, and the number of syscalls isn’t reduced).